prefect_ray/task_runners.py

```diff
@@ -183,6 +183,11 @@ async def _start(self, exit_stack: AsyncExitStack):
         - Creates a client to connect to the cluster.
         - Pushes a call to wait for all running futures to complete on exit.
         """
+        # If this method is called twice, like in a subflow, ray.init() yields:
+        # RuntimeError: Maybe you called ray.init twice by accident?
+        if ray.is_initialized():
+            return
+
```

Comment:
This should go as an `elif` after `if self.address` I think. We should also log that we're connecting to an already existing local Ray instance.

Separately, we can disable this error by passing something like `allow_reinit` as a kwarg. This may be a better way to preserve the existing code path which displays some nice information about the cluster.

Reply:
What happens if you attempt to connect to two separate Ray clusters, one in the parent flow and one in the subflow?

Reply:
I tried connecting to the just created cluster using "auto" but it still re-initializes and crashes.
```


```
15:40:39.598 | INFO    | prefect.engine - Created flow run 'scarlet-sawfish' for flow 'some-flow'
15:40:39.598 | INFO    | prefect.task_runner.ray - Creating a local Ray instance
2022-08-18 15:40:41,466 INFO services.py:1456 -- View the Ray dashboard at http://127.0.0.1:8265
15:40:44.329 | INFO    | prefect.task_runner.ray - Using Ray cluster with 1 nodes.
15:40:44.329 | INFO    | prefect.task_runner.ray - The Ray UI is available at 127.0.0.1:8265
15:40:45.939 | INFO    | Flow run 'scarlet-sawfish' - Created subflow run 'logical-marmoset' for flow 'my-subflow'
15:40:45.942 | INFO    | prefect.task_runner.ray - Connecting to an existing Ray instance at auto
2022-08-18 15:40:45,974 INFO worker.py:946 -- Connecting to existing Ray cluster at address: 127.0.0.1:61808
```

Traceback:
```
15:40:45.975 | ERROR   | Flow run 'logical-marmoset' - Crash detected! Execution was interrupted by an unexpected exception.
15:40:46.113 | ERROR   | Flow run 'scarlet-sawfish' - Encountered exception during execution:
Traceback (most recent call last):
  File "/Users/andrew/Applications/python/prefect/src/prefect/engine.py", line 566, in orchestrate_flow_run
    result = await run_sync(flow_call)
  File "/Users/andrew/Applications/python/prefect/src/prefect/utilities/asyncutils.py", line 56, in run_sync_in_worker_thread
    return await anyio.to_thread.run_sync(call, cancellable=True)
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/site-packages/anyio/to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/site-packages/anyio/_backends/_asyncio.py", line 867, in run
    result = context.run(func, *args)
  File "test_ray_subflow.py", line 19, in some_flow
    my_subflow()
  File "/Users/andrew/Applications/python/prefect/src/prefect/flows.py", line 384, in __call__
    return enter_flow_run_engine_from_flow_call(
  File "/Users/andrew/Applications/python/prefect/src/prefect/engine.py", line 156, in enter_flow_run_engine_from_flow_call
    return run_async_from_worker_thread(begin_run)
  File "/Users/andrew/Applications/python/prefect/src/prefect/utilities/asyncutils.py", line 136, in run_async_from_worker_thread
    return anyio.from_thread.run(call)
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/site-packages/anyio/from_thread.py", line 49, in run
    return asynclib.run_async_from_thread(func, *args)
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/site-packages/anyio/_backends/_asyncio.py", line 970, in run_async_from_thread
    return f.result()
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/concurrent/futures/_base.py", line 444, in result
    return self.__get_result()
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/Users/andrew/Applications/python/prefect/src/prefect/client.py", line 105, in with_injected_client
    return await fn(*args, **kwargs)
  File "/Users/andrew/Applications/python/prefect/src/prefect/engine.py", line 460, in create_and_begin_subflow_run
    task_runner = await stack.enter_async_context(flow.task_runner.start())
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/contextlib.py", line 568, in enter_async_context
    result = await _cm_type.__aenter__(cm)
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/contextlib.py", line 171, in __aenter__
    return await self.gen.__anext__()
  File "/Users/andrew/Applications/python/prefect/src/prefect/task_runners.py", line 168, in start
    await self._start(exit_stack)
  File "/Users/andrew/Applications/python/prefect-ray/prefect_ray/task_runners.py", line 195, in _start
    context = ray.init(*init_args, **self.init_kwargs)
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/site-packages/ray/worker.py", line 967, in init
    raise RuntimeError(
RuntimeError: Maybe you called ray.init twice by accident? This error can be suppressed by passing in 'ignore_reinit_error=True' or by calling 'ray.shutdown()' prior to 'ray.init()'.
15:40:49.189 | ERROR   | Flow run 'scarlet-sawfish' - Finished in state Failed('Flow run encountered an exception.')
Traceback (most recent call last):
  File "test_ray_subflow.py", line 23, in <module>
    some_flow()
  File "/Users/andrew/Applications/python/prefect/src/prefect/flows.py", line 384, in __call__
    return enter_flow_run_engine_from_flow_call(
  File "/Users/andrew/Applications/python/prefect/src/prefect/engine.py", line 152, in enter_flow_run_engine_from_flow_call
    return anyio.run(begin_run)
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/site-packages/anyio/_core/_eventloop.py", line 70, in run
    return asynclib.run(func, *args, **backend_options)
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/site-packages/anyio/_backends/_asyncio.py", line 292, in run
    return native_run(wrapper(), debug=debug)
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/site-packages/anyio/_backends/_asyncio.py", line 287, in wrapper
    return await func(*args)
  File "/Users/andrew/Applications/python/prefect/src/prefect/client.py", line 105, in with_injected_client
    return await fn(*args, **kwargs)
  File "/Users/andrew/Applications/python/prefect/src/prefect/engine.py", line 228, in create_then_begin_flow_run
    return state.result()
  File "/Users/andrew/Applications/python/prefect/src/prefect/orion/schemas/states.py", line 145, in result
    raise data
  File "/Users/andrew/Applications/python/prefect/src/prefect/engine.py", line 566, in orchestrate_flow_run
    result = await run_sync(flow_call)
  File "/Users/andrew/Applications/python/prefect/src/prefect/utilities/asyncutils.py", line 56, in run_sync_in_worker_thread
    return await anyio.to_thread.run_sync(call, cancellable=True)
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/site-packages/anyio/to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/site-packages/anyio/_backends/_asyncio.py", line 867, in run
    result = context.run(func, *args)
  File "test_ray_subflow.py", line 19, in some_flow
    my_subflow()
  File "/Users/andrew/Applications/python/prefect/src/prefect/flows.py", line 384, in __call__
    return enter_flow_run_engine_from_flow_call(
  File "/Users/andrew/Applications/python/prefect/src/prefect/engine.py", line 156, in enter_flow_run_engine_from_flow_call
    return run_async_from_worker_thread(begin_run)
  File "/Users/andrew/Applications/python/prefect/src/prefect/utilities/asyncutils.py", line 136, in run_async_from_worker_thread
    return anyio.from_thread.run(call)
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/site-packages/anyio/from_thread.py", line 49, in run
    return asynclib.run_async_from_thread(func, *args)
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/site-packages/anyio/_backends/_asyncio.py", line 970, in run_async_from_thread
    return f.result()
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/concurrent/futures/_base.py", line 444, in result
    return self.__get_result()
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/Users/andrew/Applications/python/prefect/src/prefect/client.py", line 105, in with_injected_client
    return await fn(*args, **kwargs)
  File "/Users/andrew/Applications/python/prefect/src/prefect/engine.py", line 460, in create_and_begin_subflow_run
    task_runner = await stack.enter_async_context(flow.task_runner.start())
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/contextlib.py", line 568, in enter_async_context
    result = await _cm_type.__aenter__(cm)
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/contextlib.py", line 171, in __aenter__
    return await self.gen.__anext__()
  File "/Users/andrew/Applications/python/prefect/src/prefect/task_runners.py", line 168, in start
    await self._start(exit_stack)
  File "/Users/andrew/Applications/python/prefect-ray/prefect_ray/task_runners.py", line 195, in _start
    context = ray.init(*init_args, **self.init_kwargs)
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/Users/andrew/mambaforge/envs/anyscale/lib/python3.8/site-packages/ray/worker.py", line 967, in init
    raise RuntimeError(
RuntimeError: Maybe you called ray.init twice by accident? This error can be suppressed by passing in 'ignore_reinit_error=True' or by calling 'ray.shutdown()' prior to 'ray.init()'
```

```
from prefect import task, flow
from prefect_ray import RayTaskRunner


@task
def some_task(input):
    return input


@flow(task_runner=RayTaskRunner(address="auto"))
def my_subflow():
    some_task.submit(1)
    some_task.submit(2)
    some_task.submit(3)


@flow(task_runner=RayTaskRunner())
def some_flow():
    my_subflow()
    some_task.submit(4)


some_flow()
```

Reply:
Is it possible to have the same Python process connect to more than one Ray cluster at once?

Reply:
The `"auto"` address is a special case since `ray` will attempt to create a cluster. I found that this code worked with your example @ahuang11:
```python
if self.address and self.address != "auto":
    self.logger.info(
        f"Connecting to an existing Ray instance at {self.address}"
    )
    init_args = (self.address,)
elif ray.is_initialized():
    self.logger.info(
        f"Local Ray instance is already initialized. Using existing local instance."
    )
    return
else:
    self.logger.info("Creating a local Ray instance")
    init_args = ()
```

Reply:
Yeah that makes more sense to me